Kernel/Bus/USB/Drivers/HID/KeyboardDriver.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2023, Liav A. <liavalb@hotmail.co.il>
+ * Copyright (c) 2024, Olekoop <mlglol360xd@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Find.h>
+#include <Kernel/Bus/USB/Drivers/HID/Codes.h>
+#include <Kernel/Bus/USB/Drivers/HID/KeyboardDriver.h>
+#include <Kernel/Bus/USB/USBClasses.h>
+#include <Kernel/Bus/USB/USBEndpoint.h>
+#include <Kernel/Bus/USB/USBRequest.h>
+#include <Kernel/Devices/DeviceManagement.h>
+#include <Kernel/Devices/HID/KeyboardDevice.h>
+#include <Kernel/Devices/HID/Management.h>
+#include <Kernel/Devices/HID/USB/KeyboardDevice.h>
+
+namespace Kernel::USB {
+
+USB_DEVICE_DRIVER(KeyboardDriver);
+
+void KeyboardDriver::init()
+{
+    auto driver = MUST(adopt_nonnull_lock_ref_or_enomem(new KeyboardDriver()));
+    USBManagement::register_driver(driver);
+}
+
+ErrorOr<void> KeyboardDriver::checkout_interface(USB::Device& device, USBInterface const& interface)
+{
+    auto const& descriptor = interface.descriptor();
+    if (descriptor.interface_class_code == USB_CLASS_HID
+        && descriptor.interface_sub_class_code == to_underlying(HID::SubclassCode::BootProtocol)
+        && descriptor.interface_protocol == to_underlying(HID::InterfaceProtocol::Keyboard)) {
+        dmesgln("USB HID Keyboard Interface for device {:#04x}:{:#04x} found", device.device_descriptor().vendor_id, device.device_descriptor().product_id);
+        return initialize_device(device, interface);
+    }
+    return ENOTSUP;
+}
+
+ErrorOr<void> KeyboardDriver::probe(USB::Device& device)
+{
+    if (device.device_descriptor().device_class != USB_CLASS_DEVICE
+        || device.device_descriptor().device_sub_class != 0x00
+        || device.device_descriptor().device_protocol != 0x00)
+        return ENOTSUP;
+    // FIXME: Are we guaranteed to have one USB configuration for a keyboard device?
+    if (device.configurations().size() != 1)
+        return ENOTSUP;
+    // FIXME: If we have multiple USB configurations for a keyboard device, find the appropriate one
+    // and handle multiple interfaces for it.
+    if (device.configurations()[0].interfaces().size() != 1)
+        return ENOTSUP;
+
+    TRY(checkout_interface(device, device.configurations()[0].interfaces()[0]));
+
+    return ENOTSUP;
+}
+
+ErrorOr<void> KeyboardDriver::initialize_device(USB::Device& device, USBInterface const& interface)
+{
+    if (interface.endpoints().size() != 1)
+        return ENOTSUP;
+    auto const& configuration = interface.configuration();
+    // FIXME: Should we check other configurations?
+    TRY(device.control_transfer(
+        USB_REQUEST_RECIPIENT_DEVICE | USB_REQUEST_TYPE_STANDARD | USB_REQUEST_TRANSFER_DIRECTION_HOST_TO_DEVICE,
+        USB_REQUEST_SET_CONFIGURATION, configuration.configuration_id(), 0, 0, nullptr));
+
+    auto const& endpoint_descriptor = interface.endpoints()[0];
```

Comment:
Are we guaranteed to have the mouse device at endpoint 0?

Reply:
I think you mean a keyboard device but yes.
The driver checks if the device is a keyboard and if it uses Boot Protocol.
Also endpoint 0 is always a control endpoint and we are guaranteed that it is available.